### ISSUE TITLE
cuda.grid(1) returns incorrect value due to signed integer overflow

### ISSUE THREAD

**Issue by nathanfarlow** [numba/numba#9229]
`cuda.grid(1)` returns an incorrect value in the following code:

```py
import numpy as np
from numba import cuda


@cuda.jit
def bug(out):
    i = cuda.threadIdx.x + cuda.blockIdx.x * cuda.blockDim.x
    i2 = cuda.grid(1)
    if i != i2:
        # this shouldn't happen
        print(i, i2)


free_mem, _ = cuda.current_context().get_memory_info()
# max_num_parallel_iter is around 566281830 on my 4090
max_num_parallel_iter = free_mem // 4 - 128 * 1024 * 1024
dev_out = cuda.device_array(max_num_parallel_iter, dtype=np.float32)
bug[22121216, 256](dev_out)
```

One example of a case which is printed is when `i = 2147551327` and `i2 = -214741596`, which likely indicates a signed 32 bit integer overflow. The bug doesn't appear when setting `max_num_parallel_iter` to an integer constant, which must mean it's somehow related to the calculation from the memory info.

**Comment by gmarkall** [comment]
Thanks for the report! I can reproduce this, and it is due to an integer overflow.

> The bug doesn't appear when setting `max_num_parallel_iter` to an integer constant, which must mean it's somehow related to the calculation from the memory info.

I think the reason you don't see it is that launches are asynchronous, and setting `max_num_parallel_iter` to a constant changed the timing of execution such that your program manages to exit before the values printed by the kernel get flushed into the terminal. I find that (and I think you should be able to reproduce too) adding `cuda.synchronize()` to the end of the program manifests the error even when there's no computation or device array:

```
from numba import cuda

@cuda.jit
def bug():
    i = cuda.threadIdx.x + cuda.blockIdx.x * cuda.blockDim.x
    i2 = cuda.grid(1)
    if i != i2:
        # this shouldn't happen
        print(i, i2)

bug[22121216, 256]()
cuda.synchronize()
```

**Comment by gmarkall** [comment]
I have a fix in https://github.com/gmarkall/numba/commit/c40f740c2f9c0831ac1a7d6fada919ce60605bc7 - I will need to add tests to this for a PR, but intend to do this soon.

**Comment by gmarkall** [comment]
@nathanfarlow There is a fix in PR #9235. As a workaround for now, can you write your kernels such that they don't require more than 2^31 threads, or use the manual computation?

One reason what seems like an "obvious" problem has slipped by for so long is that I think people generally don't use so many threads,  so this isn't a common pattern in practice. Instead, it's usually more efficient to use a grid-strided loop with fewer threads, as outlined in: https://towardsdatascience.com/cuda-by-numba-examples-1-4-e0d06651612f#0909

**Comment by nathanfarlow** [comment]
Thanks for looking into this and thanks for the advice! :)
